File: internal/ocr/pdf.go
package ocr

import (
	"bytes"
	"fmt"
	"image"
	"image/color"
	"image/draw"
	"io"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/disintegration/imaging"
	"github.com/ledongthuc/pdf"
	"golang.org/x/image/font"
	"golang.org/x/image/font/basicfont"
	"golang.org/x/image/math/fixed"
)

const (
	minPDFLetters   = 40
	maxPDFPages     = 40
	maxPDFTextRunes = 80_000
	previewPageGap  = 12
	previewMaxW     = 1200
)

var previewGapColor = color.RGBA{R: 216, G: 222, B: 230, A: 255}

func extractPDFText(raw []byte) (text string, err error) {
	defer func() {
		if rec := recover(); rec != nil {
			text = ""
			err = nil
		}
	}()
	raw = pdfPayload(raw)
	r, err := pdf.NewReader(bytes.NewReader(raw), int64(len(raw)))
	if err != nil {
		return "", nil
	}
	n := r.NumPage()
	if n > maxPDFPages {
		n = maxPDFPages
	}
	var b strings.Builder
	for i := 1; i <= n; i++ {
		page := pageText(r.Page(i))
		if page == "" {
			continue
		}
		if b.Len() > 0 {
			b.WriteString("\n\n")
		}
		b.WriteString(page)
	}
	out := strings.TrimSpace(b.String())
	if out == "" {
		rd, plainErr := r.GetPlainText()
		if plainErr == nil && rd != nil {
			var buf bytes.Buffer
			_, _ = io.Copy(&buf, rd)
			out = strings.TrimSpace(buf.String())
		}
	}
	if utf8.RuneCountInString(out) > maxPDFTextRunes {
		out = string([]rune(out)[:maxPDFTextRunes])
	}
	return out, nil
}

func pageText(p pdf.Page) string {
	if p.V.IsNull() {
		return ""
	}
	rows, err := p.GetTextByRow()
	if err != nil {
		return ""
	}
	var b strings.Builder
	for _, row := range rows {
		var parts []string
		for _, word := range row.Content {
			s := strings.TrimSpace(word.S)
			if s != "" {
				parts = append(parts, s)
			}
		}
		if len(parts) == 0 {
			continue
		}
		b.WriteString(strings.Join(parts, " "))
		b.WriteByte('\n')
	}
	return strings.TrimSpace(b.String())
}

func isPDFWithText(text string) bool {
	n := 0
	for _, r := range text {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			n++
			if n >= minPDFLetters {
				return true
			}
		}
	}
	return false
}

func pdfPayload(raw []byte) []byte {
	i := bytes.Index(raw, []byte("%PDF-"))
	if i <= 0 {
		return raw
	}
	return raw[i:]
}

func previewPDF(raw []byte) ([]byte, error) {
	if jpeg, err := previewPDFPages(raw); err == nil && len(jpeg) > 0 {
		return jpeg, nil
	}
	text, _ := extractPDFText(raw)
	return renderTextSlip(text)
}

func previewPDFPages(raw []byte) ([]byte, error) {
	pngs, err := previewPagePNGs(raw)
	if err != nil {
		return nil, err
	}
	return stackPreviewJPEG(pngs)
}

func stackPreviewJPEG(pngs [][]byte) ([]byte, error) {
	if len(pngs) == 0 {
		return nil, fmt.Errorf("could not rasterize the PDF")
	}
	pages := make([]image.Image, 0, len(pngs))
	for _, raw := range pngs {
		img, err := imaging.Decode(bytes.NewReader(raw), imaging.AutoOrientation(true))
		if err != nil {
			return nil, fmt.Errorf("could not read the PDF page")
		}
		pages = append(pages, fitPreviewPage(flattenWhite(img)))
	}
	return fitJPEG(stackPages(pages), maxPreparedBytes)
}

func fitPreviewPage(img image.Image) image.Image {
	if img.Bounds().Dx() > previewMaxW {
		return imaging.Resize(img, previewMaxW, 0, imaging.Lanczos)
	}
	return img
}

func stackPages(pages []image.Image) image.Image {
	if len(pages) == 0 {
		return imaging.New(1, 1, color.White)
	}
	if len(pages) == 1 {
		return pages[0]
	}
	width := 0
	height := 0
	for i, p := range pages {
		if w := p.Bounds().Dx(); w > width {
			width = w
		}
		height += p.Bounds().Dy()
		if i > 0 {
			height += previewPageGap
		}
	}
	out := imaging.New(width, height, previewGapColor)
	y := 0
	for i, p := range pages {
		x := (width - p.Bounds().Dx()) / 2
		out = imaging.Paste(out, p, image.Pt(x, y))
		y += p.Bounds().Dy()
		if i < len(pages)-1 {
			y += previewPageGap
		}
	}
	return out
}

func renderTextSlip(text string) ([]byte, error) {
	lines := wrapPreviewLines(text)
	face := basicfont.Face7x13
	const (
		pad   = 18
		lineH = 14
		width = 420
	)
	height := pad*2 + lineH*len(lines)
	if height < 220 {
		height = 220
	}
	img := image.NewRGBA(image.Rect(0, 0, width, height))
	draw.Draw(img, img.Bounds(), &image.Uniform{color.White}, image.Point{}, draw.Src)
	d := &font.Drawer{
		Dst:  img,
		Src:  image.NewUniform(color.RGBA{R: 16, G: 32, B: 51, A: 255}),
		Face: face,
		Dot:  fixed.P(pad, pad+11),
	}
	for _, line := range lines {
		d.DrawString(line)
		d.Dot.X = fixed.I(pad)
		d.Dot.Y += fixed.I(lineH)
	}
	return encodeJPEG(img)
}

func wrapPreviewLines(text string) []string {
	text = strings.TrimSpace(text)
	if text == "" {
		return []string{"PDF bill", "Could not render PDF pages."}
	}
	const (
		maxChars = 52
		maxLines = 72
	)
	var out []string
	out = append(out, "PDF bill")
	out = append(out, "")
	for _, raw := range strings.Split(text, "\n") {
		raw = strings.TrimSpace(raw)
		if raw == "" {
			if len(out) > 0 && out[len(out)-1] != "" {
				out = append(out, "")
			}
			continue
		}
		runes := []rune(raw)
		for len(runes) > 0 {
			n := maxChars
			if n > len(runes) {
				n = len(runes)
			}
			out = append(out, string(runes[:n]))
			runes = runes[n:]
			if len(out) >= maxLines {
				return out
			}
		}
		if len(out) >= maxLines {
			return out
		}
	}
	return out
}

File: src/ocr/pdf.ts
import { execFile } from 'node:child_process';
import { constants } from 'node:fs';
import { access, mkdtemp, readdir, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { basename, delimiter, extname, join } from 'node:path';
import { promisify } from 'node:util';
import sharp from 'sharp';
import { flattenWhite, fitJPEG, maxPreparedBytes, prepareJPEG } from './image.js';
import { NoPDFTextError } from './types.js';

const execFileAsync = promisify(execFile);

export const minPDFLetters = 40;
export const maxPDFPages = 40;
const maxPDFTextRunes = 80_000;
const previewPageGap = 12;
const previewMaxW = 1200;
const visionDPI = '200';
const previewDPI = '150';
const previewGapColor = { r: 216, g: 222, b: 230 };

export async function previewPDF(raw: Buffer): Promise<Buffer> {
  try {
    const jpeg = await previewPDFPages(raw);
    if (jpeg.length > 0) return jpeg;
  } catch {
    /* fall through to text slip */
  }
  const text = await extractPDFText(raw);
  return renderTextSlip(text);
}

export async function pdfPageJPEGs(raw: Buffer): Promise<Buffer[]> {
  const pages = await withRasterized(raw, visionDPI, async (paths) => {
    const out: Buffer[] = [];
    for (const path of paths) {
      const b = await readFile(path);
      if (b.length === 0) continue;
      out.push(await prepareJPEG(b));
    }
    return out;
  });
  if (pages.length === 0) throw new NoPDFTextError();
  return pages;
}

export async function extractPDFText(raw: Buffer): Promise<string> {
  try {
    const dir = await mkdtemp(join(tmpdir(), 'bulkly-pdf-'));
    try {
      const pdfPath = join(dir, 'in.pdf');
      await writeFile(pdfPath, pdfPayload(raw), { mode: 0o600 });
      const { stdout } = await execFileAsync('pdftotext', ['-layout', '-l', String(maxPDFPages), pdfPath, '-'], {
        timeout: 120_000,
        maxBuffer: 8 << 20,
      });
      let out = stdout.trim();
      const runes = [...out];
      if (runes.length > maxPDFTextRunes) out = runes.slice(0, maxPDFTextRunes).join('');
      return out;
    } finally {
      await rm(dir, { recursive: true, force: true });
    }
  } catch {
    return '';
  }
}

export function isPDFWithText(text: string): boolean {
  let n = 0;
  for (const r of text) {
    if (/[\p{L}\p{N}]/u.test(r)) {
      n++;
      if (n >= minPDFLetters) return true;
    }
  }
  return false;
}

export function pdfPayload(raw: Buffer): Buffer {
  const i = raw.indexOf('%PDF-');
  if (i <= 0) return raw;
  return raw.subarray(i);
}

async function previewPDFPages(raw: Buffer): Promise<Buffer> {
  const pngs = await previewPagePNGs(raw);
  return stackPreviewJPEG(pngs);
}

async function previewPagePNGs(raw: Buffer): Promise<Buffer[]> {
  return withRasterized(raw, previewDPI, async (paths) => {
    const out: Buffer[] = [];
    for (const path of paths) {
      const b = await readFile(path);
      if (b.length === 0) continue;
      out.push(b);
    }
    if (out.length === 0) throw new Error('could not rasterize the PDF');
    return out;
  });
}

export async function stackPreviewJPEG(pngs: Buffer[]): Promise<Buffer> {
  if (pngs.length === 0) throw new Error('could not rasterize the PDF');
  const pages: Buffer[] = [];
  for (const raw of pngs) {
    let img: Buffer;
    try {
      img = await flattenWhite(raw);
    } catch {
      throw new Error('could not read the PDF page');
    }
    pages.push(await fitPreviewPage(img));
  }
  return fitJPEG(await stackPages(pages), maxPreparedBytes);
}

async function fitPreviewPage(img: Buffer): Promise<Buffer> {
  const meta = await sharp(img).metadata();
  if ((meta.width ?? 0) > previewMaxW) {
    return sharp(img).resize({ width: previewMaxW, kernel: 'lanczos3' }).png().toBuffer();
  }
  return img;
}

async function stackPages(pages: Buffer[]): Promise<Buffer> {
  if (pages.length === 0) {
    return sharp({ create: { width: 1, height: 1, channels: 3, background: '#ffffff' } })
      .png()
      .toBuffer();
  }
  if (pages.length === 1) return pages[0]!;
  const metas = await Promise.all(pages.map((p) => sharp(p).metadata()));
  let width = 0;
  let height = 0;
  for (let i = 0; i < metas.length; i++) {
    const w = metas[i]!.width ?? 1;
    const h = metas[i]!.height ?? 1;
    if (w > width) width = w;
    height += h;
    if (i > 0) height += previewPageGap;
  }
  const composites: sharp.OverlayOptions[] = [];
  let y = 0;
  for (let i = 0; i < pages.length; i++) {
    const w = metas[i]!.width ?? 1;
    const h = metas[i]!.height ?? 1;
    const x = Math.floor((width - w) / 2);
    composites.push({ input: pages[i], top: y, left: x });
    y += h;
    if (i < pages.length - 1) y += previewPageGap;
  }
  return sharp({
    create: { width, height, channels: 3, background: previewGapColor },
  })
    .composite(composites)
    .png()
    .toBuffer();
}

export async function renderTextSlip(text: string): Promise<Buffer> {
  const lines = wrapPreviewLines(text);
  const pad = 18;
  const lineH = 14;
  const width = 420;
  let height = pad * 2 + lineH * lines.length;
  if (height < 220) height = 220;
  const texts = lines
    .map((line, i) => {
      const y = pad + 11 + i * lineH;
      return `<text x="${pad}" y="${y}" xml:space="preserve" font-family="ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, monospace" font-size="13" fill="rgb(16,32,51)">${escapeXml(line)}</text>`;
    })
    .join('');
  const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}"><rect width="100%" height="100%" fill="white"/>${texts}</svg>`;
  return sharp(Buffer.from(svg)).jpeg({ quality: 82 }).toBuffer();
}

export function wrapPreviewLines(text: string): string[] {
  text = text.trim();
  if (text === '') return ['PDF bill', 'Could not render PDF pages.'];
  const maxChars = 52;
  const maxLines = 72;
  const out: string[] = ['PDF bill', ''];
  for (let raw of text.split('\n')) {
    raw = raw.trim();
    if (raw === '') {
      if (out.length > 0 && out[out.length - 1] !== '') out.push('');
      continue;
    }
    let runes = [...raw];
    while (runes.length > 0) {
      const n = Math.min(maxChars, runes.length);
      out.push(runes.slice(0, n).join(''));
      runes = runes.slice(n);
      if (out.length >= maxLines) return out;
    }
    if (out.length >= maxLines) return out;
  }
  return out;
}

function escapeXml(s: string): string {
  return s.replaceAll('&', '&amp;').replaceAll('<', '&lt;').replaceAll('>', '&gt;').replaceAll('"', '&quot;');
}

function errNeedPoppler(): NoPDFTextError {
  return new NoPDFTextError('this PDF could not be read as images: install poppler or run Bulkly in Docker');
}

async function withRasterized<T>(raw: Buffer, dpi: string, fn: (paths: string[]) => Promise<T>): Promise<T> {
  if (!(await commandExists('pdftoppm'))) throw errNeedPoppler();
  const { pages, cleanup } = await rasterizePDF(raw, dpi, maxPDFPages);
  try {
    if (pages.length === 0) throw new Error('could not rasterize the PDF');
    return await fn(pages);
  } finally {
    await cleanup();
  }
}

async function rasterizePDF(
  raw: Buffer,
  dpi: string,
  lastPage: number,
): Promise<{ pages: string[]; cleanup: () => Promise<void> }> {
  const dir = await mkdtemp(join(tmpdir(), 'bulkly-pdf-'));
  const cleanup = async () => {
    await rm(dir, { recursive: true, force: true });
  };
  try {
    const pdfPath = join(dir, 'in.pdf');
    await writeFile(pdfPath, pdfPayload(raw), { mode: 0o600 });
    const prefix = join(dir, 'page');
    try {
      await execFileAsync('pdftoppm', ['-png', '-r', dpi, '-l', String(lastPage), pdfPath, prefix], {
        timeout: 120_000,
        maxBuffer: 16 << 20,
      });
    } catch (err) {
      if (isEnoent(err)) throw errNeedPoppler();
      const msg = stderrMessage(err);
      throw new Error(msg ? `could not rasterize the PDF: pdftoppm: ${msg}` : `could not rasterize the PDF: ${String(err)}`);
    }
    const names = await readdir(dir);
    const matches = names.filter((n) => n.startsWith('page-') && n.endsWith('.png')).map((n) => join(dir, n));
    sortPageFiles(matches);
    return { pages: matches, cleanup };
  } catch (err) {
    await cleanup();
    throw err;
  }
}

function sortPageFiles(files: string[]): void {
  files.sort((a, b) => pageIndex(a) - pageIndex(b));
}

function pageIndex(path: string): number {
  const base = basename(path, extname(path));
  const i = base.lastIndexOf('-');
  if (i < 0) return 0;
  const n = Number.parseInt(base.slice(i + 1), 10);
  return Number.isFinite(n) ? n : 0;
}

async function commandExists(cmd: string): Promise<boolean> {
  for (const dir of (process.env.PATH ?? '').split(delimiter)) {
    if (!dir) continue;
    try {
      await access(join(dir, cmd), constants.X_OK);
      return true;
    } catch {
      /* continue */
    }
  }
  return false;
}

function isEnoent(err: unknown): boolean {
  return Boolean(err && typeof err === 'object' && 'code' in err && (err as { code?: string }).code === 'ENOENT');
}

function stderrMessage(err: unknown): string {
  if (err && typeof err === 'object' && 'stderr' in err) {
    const stderr = (err as { stderr?: Buffer | string }).stderr;
    return (typeof stderr === 'string' ? stderr : stderr?.toString('utf8') ?? '').trim();
  }
  return err instanceof Error ? err.message : '';
}
